### ISSUE TITLE
Discuss: handling form input

### ISSUE THREAD

**Issue by caolan** [caolan/magery#1]
**Controlled**

Currently, all elements are considered 'controlled' (in the React sense). That means, to update a form element you _must_ intercept `oninput` or similar events then update the data and re-patch the page. This gives the most control over these elements using only template data. However, it is also quite invasive as it effectively disables all inputs and instead requires re-patching on every keypress or button click. This is the only way to reliably control the real (displayed) value on a text input using only template data.

``` html
{{#define example}}
  <input type="text" value="{{description}}" oninput="updateDescription" />
  <button onclick="save">Save</button>
{{/define}}
```

``` javascript
container.dispatch = function (name, event, context) {
    if (name == 'updateDescription') {
        context.description = event.target.value;
    }
    if (name == 'save') {
        saveData(context.description);
    }
    patch();
};
```

**Uncontrolled**

The alternative is to leave all elements uncontrolled. That means we do not intercept input events or force the user to re-patch to update them. When uncontrolled, the form elements would become a secondary store of state in the application (as their value property may differ from their value attribute). You would not be able to update the value property using template data. So effectively you can't change the live/displayed value using only template data and patching. _However_, this is less invasive and less likely to subvert normal browser behaviour (e.g. clearing a textbox after a user types data into an input in the gap between page load and javascript running). For this to be workable we'd need a way to refer to elements on the page without traversing the DOM. This gives us a light touch with the browser, which is desirable, and I'd be interested in your ideas on how this could work.

``` html
{{#define example}}
  <form>
    <input type="text" name="desc" value="{{description}}" />
    <button onclick="save">Save</button>
  </form>
{{/define}}
```

``` javascript
container.dispatch = function (name, event, context) {
    if (name == 'save') {
        saveData(event.target.form.desc.value);
    }
    patch();
};
```

**Hybrid**

The third option is to try and do both. React allows you to choose between these approaches depending on whether a value attribute is present. We can't do this easily since some form elements (e.g. checkbox) use a boolean attribute to set their value. Omitting that attribute means it may be unchecked, for example. We can't allow assigning `checked={false}` as when rendered server-side it would be interpreted differently. So we need to either always have controlled elements, always have uncontrolled elements, or introduce some custom attributes which let us choose. I'd like to keep element customization to a minimum where possible so I'm not as keen on this approach.

**Notes**

Whatever we choose needs to make sense in a server context too. Adding a custom attribute that sets the checked property on a form element might require waiting for JavaScript to load before it displays correctly. If possible, the initial server-side render should exactly match the initial JS patch.

**Discuss**

I'd like your feedback on which approach makes sense. So far the _everything is controlled_ approach works, but is invasive. I want to hear your ideas on how we might improve this.

**Comment by glenjamin** [comment]
Based on my React experiences, hybrid is useful enough to try and achieve.

One option would be to use the `defaultValue` that React has as the sentinel.

I don't think the server interpretation of `checked={false}` is a problem, because a server rendered page can't be "controlled". When the JS took over it would become controlled.

**Comment by caolan** [comment]
But until the JS takes over the browser would interpret checked=false as being checked. So it'd have to be on some other property, e.g. data-checked

**Comment by glenjamin** [comment]
> But until the JS takes over the browser would interpret checked=false as being checked. So it'd have to be on some other property, e.g. data-checked

I was imagining that `<input type="checkbox" checked={false} />` would render as `<input type="checkbox" />` in server mode. I don't think you need to know the user had it as an explicit false or not in the server case.

**Comment by caolan** [comment]
The server rendering is currently very simple and it doesn't require the server to understand or parse the markup itself. It can just do string interpolation like a regular templating language. So conditional rendering of attributes wouldn't work so well.

**Comment by caolan** [comment]
The syntax has changed quite a bit since we had this discussion, but see https://github.com/caolan/magery#data-managed for an explanation of the current solution to this problem.

tl;dr: defaults to unmanaged, but managed if the `data-managed="true"` is present.
